File: problems/2415. Reverse Odd Levels of Binary Tree/cpp/main.cpp
#include <bits/stdc++.h>
using namespace std;

#define sz(s) (int)s.size()


struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right)
        : val(x), left(left), right(right) {}
};

TreeNode *tn(vector<int> &v, int idx = 1) {
    if (idx > sz(v))
        return nullptr;

    TreeNode *t = new TreeNode(v[idx - 1]);
    t->left = tn(v, idx * 2);
    t->right = tn(v, idx * 2 + 1);

    return t;
}

vector<int> vec(TreeNode *t) {
    queue<TreeNode *> ts;
    ts.push(t);
    vector<int> out;

    while (!ts.empty()) {
        auto curr = ts.front();
        ts.pop();
        if (curr == nullptr)
            continue;

        ts.push(curr->left);
        ts.push(curr->right);
        out.push_back(curr->val);
    }

    return out;
}

class Solution {
  public:
    TreeNode *reverseOddLevels(TreeNode *root) {
        vector<TreeNode *> evens, odds;
        evens.push_back(root);

        while (!evens.empty() || !odds.empty()) {
            for (int i = 0; i < sz(evens); i++) {
                auto curr = evens[i];

                if (curr->left)
                    odds.push_back(curr->left);
                if (curr->right)
                    odds.push_back(curr->right);
            }

			evens.clear();

            for (int i = 0; i < sz(odds); i++) {
                auto curr = odds[i];

				if (i < sz(odds) / 2) {
					auto tmp = curr->val;
					int j = sz(odds) - i - 1;

					curr->val = odds[j]->val;
					odds[j]->val = tmp;
				}

                if (curr->left)
                    evens.push_back(curr->left);
                if (curr->right)
                    evens.push_back(curr->right);
            }

			odds.clear();
        }

        return root;
    }
};


int main() {
    auto s = Solution();

    vector<int> v, exp;
    TreeNode *t;

    v = {2, 3, 5, 8, 13, 21, 34};
    t = tn(v);
    exp = {2, 5, 3, 8, 13, 21, 34};

    assert(vec(s.reverseOddLevels(t)) == exp);
}

File: problems/2415. Reverse Odd Levels of Binary Tree/python/main.py
from typing import Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def tn(lis: list, idx=1):
    if idx > len(lis) or lis[idx - 1] is None:
        return None

    t = TreeNode(lis[idx - 1])
    t.left = tn(lis, idx * 2)
    t.right = tn(lis, idx * 2 + 1)

    return t


def lis(t: Optional[TreeNode]):
    l = []
    nodes = [t]

    while nodes:
        n = nodes.pop(0)
        if n is None:
            l.append(None)
            continue

        l.append(n.val)

        nodes.append(n.left)
        nodes.append(n.right)

    for i in range(len(l) - 1, -1, -1):
        if l[i] is not None:
            return l[: i + 1]

    return l


class Solution:
    def reverseOddLevels(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        nodes = [(0, root)]
        levels = {}

        # first run, put values in list
        while nodes:
            lvl, n = nodes.pop(0)
            if n is None:
                continue

            if lvl % 2 == 1:
                if lvl not in levels:
                    levels[lvl] = []
                levels[lvl].append(n.val)

            nodes.append((lvl + 1, n.left))
            nodes.append((lvl + 1, n.right))

        nodes = [(0, root)]
        while nodes:
            lvl, n = nodes.pop(0)

            if n is None:
                continue

            if lvl % 2 == 1:
                n.val = levels[lvl].pop()

            nodes.append((lvl + 1, n.left))
            nodes.append((lvl + 1, n.right))

        return root


if __name__ == "__main__":
    org = [1, 2, 3, 4]
    new = lis(tn(org))
    assert org == new

    s = Solution()
    org = [2, 3, 5, 8, 13, 21, 34]
    got = lis(s.reverseOddLevels(tn(org)))
    exp = [
        2,
        5,
        3,
        8,
        13,
        21,
        34,
    ]

    assert got == exp
